### PR TITLE
Fix a bug where the KubeletConfiguration is not printed correctly in the log

### PR DESCRIPTION
**Pull request by SataQiu**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Fix a bug where the `KubeletConfiguration` is not printed correctly in the log.

Without this patch, only the TypeMeta information can be printed, the output like this:
```
I0222 09:03:38.494769    1711 server.go:267] "KubeletConfiguration" configuration="&TypeMeta{Kind:,APIVersion:,}"
```
This is because `KubeletConfiguration` does not implement `String()` method, and the `String()` method of the nested `metav1.TypeMeta` will be called incorrectly.

With this patch:
```sh
I0510 14:24:12.992457     800 server.go:260] "KubeletConfiguration" configuration=<
        {
          "EnableServer": true,
          "StaticPodPath": "/etc/kubernetes/manifests",
          "SyncFrequency": "1m0s",
          "FileCheckFrequency": "20s",
          "HTTPCheckFrequency": "20s",
          "StaticPodURL": "",
          "StaticPodURLHeader": null,
          "Address": "0.0.0.0",
          "Port": 10250,
          "ReadOnlyPort": 0,
          "VolumePluginDir": "/usr/libexec/kubernetes/kubelet-plugins/volume/exec/",
          "ProviderID": "kind://docker/kind/kind-worker",
          "TLSCertFile": "/var/lib/kubelet/pki/kubelet.crt",
          "TLSPrivateKeyFile": "/var/lib/kubelet/pki/kubelet.key",
          "TLSCipherSuites": null,
          "TLSMinVersion": "",
          "RotateCertificates": true,
          "ServerTLSBootstrap": false,
          "Authentication": {
            "X509": {
              "ClientCAFile": "/etc/kubernetes/pki/ca.crt"
            },
            "Webhook": {
              "Enabled": true,
              "CacheTTL": "2m0s"
            },
            "Anonymous": {
              "Enabled": false
            }
          },
          "Authorization": {
            "Mode": "Webhook",
            "Webhook": {
              "CacheAuthorizedTTL": "5m0s",
              "CacheUnauthorizedTTL": "30s"
            }
          },
          "RegistryPullQPS": 5,
          "RegistryBurst": 10,
          "EventRecordQPS": 50,
          "EventBurst": 100,
          "EnableDebuggingHandlers": true,
          "EnableContentionProfiling": false,
          "HealthzPort": 10248,
          "HealthzBindAddress": "127.0.0.1",
          "OOMScoreAdj": -999,
          "ClusterDomain": "cluster.local",
          "ClusterDNS": [
            "10.96.0.10"
          ],
          "StreamingConnectionIdleTimeout": "4h0m0s",
          "NodeStatusUpdateFrequency": "10s",
          "NodeStatusReportFrequency": "5m0s",
          "NodeLeaseDurationSeconds": 40,
          "ImageMinimumGCAge": "2m0s",
          "ImageGCHighThresholdPercent": 100,
          "ImageGCLowThresholdPercent": 80,
          "VolumeStatsAggPeriod": "1m0s",
          "KubeletCgroups": "",
          "SystemCgroups": "",
          "CgroupRoot": "/kubelet",
          "CgroupsPerQOS": true,
          "CgroupDriver": "systemd",
          "CPUManagerPolicy": "none",
          "CPUManagerPolicyOptions": null,
          "CPUManagerReconcilePeriod": "10s",
          "MemoryManagerPolicy": "None",
          "TopologyManagerPolicy": "none",
          "TopologyManagerScope": "container",
          "TopologyManagerPolicyOptions": null,
          "QOSReserved": null,
          "RuntimeRequestTimeout": "2m0s",
          "HairpinMode": "promiscuous-bridge",
          "MaxPods": 110,
          "PodCIDR": "",
          "PodPidsLimit": -1,
          "ResolverConfig": "/etc/resolv.conf",
          "RunOnce": false,
          "CPUCFSQuota": true,
          "CPUCFSQuotaPeriod": "100ms",
          "MaxOpenFiles": 1000000,
          "NodeStatusMaxImages": 50,
          "ContentType": "application/vnd.kubernetes.protobuf",
          "KubeAPIQPS": 50,
          "KubeAPIBurst": 100,
          "SerializeImagePulls": true,
          "MaxParallelImagePulls": null,
          "EvictionHard": {
            "imagefs.available": "0%",
            "nodefs.available": "0%",
            "nodefs.inodesFree": "0%"
          },
          "EvictionSoft": null,
          "EvictionSoftGracePeriod": null,
          "EvictionPressureTransitionPeriod": "5m0s",
          "EvictionMaxPodGracePeriod": 0,
          "EvictionMinimumReclaim": null,
          "PodsPerCore": 0,
          "EnableControllerAttachDetach": true,
          "ProtectKernelDefaults": false,
          "MakeIPTablesUtilChains": true,
          "IPTablesMasqueradeBit": 14,
          "IPTablesDropBit": 15,
          "FeatureGates": null,
          "FailSwapOn": false,
          "MemorySwap": {
            "SwapBehavior": ""
          },
          "ContainerLogMaxSize": "10Mi",
          "ContainerLogMaxFiles": 5,
          "ConfigMapAndSecretChangeDetectionStrategy": "Watch",
          "AllowedUnsafeSysctls": null,
          "KernelMemcgNotification": false,
          "SystemReserved": null,
          "KubeReserved": null,
          "SystemReservedCgroup": "",
          "KubeReservedCgroup": "",
          "EnforceNodeAllocatable": [
            "pods"
          ],
          "ReservedSystemCPUs": "",
          "ShowHiddenMetricsForVersion": "",
          "Logging": {
            "format": "text",
            "flushFrequency": 5000000000,
            "verbosity": 5,
            "options": {
              "json": {
                "infoBufferSize": "0"
              }
            }
          },
          "EnableSystemLogHandler": true,
          "EnableSystemLogQuery": false,
          "ShutdownGracePeriod": "0s",
          "ShutdownGracePeriodCriticalPods": "0s",
          "ShutdownGracePeriodByPodPriority": null,
          "ReservedMemory": null,
          "EnableProfilingHandler": true,
          "EnableDebugFlagsHandler": true,
          "SeccompDefault": false,
          "MemoryThrottlingFactor": 0.9,
          "RegisterWithTaints": null,
          "RegisterNode": true,
          "Tracing": null,
          "LocalStorageCapacityIsolation": true,
          "ContainerRuntimeEndpoint": "unix:///run/containerd/containerd.sock",
          "ImageServiceEndpoint": ""
        }
 >
```

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by saschagrunert — APPROVED** [review]

File: cmd/kubelet/app/server.go
Comment (bart0sh) [inline cmd/kubelet/app/server.go:258]: @SataQiu Would it make sense to implement String() method for KubeletConfiguration instead?

Comment (SataQiu) [reply cmd/kubelet/app/server.go:258]: In my opinion, this usage is just a special case, and the internal configuration objects(such as `KubeletConfiguration`、`KubeproxyConfiguration`) usually don't need to implement the `String()` method. 
For the resource API objects, we've already provided `String()` method by code generation, such as:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/apps/v1/generated.pb.go#L3361

Comment (bart0sh) [reply cmd/kubelet/app/server.go:258]: Thanks for the explanations. Makes sense to me.

**Comment by SataQiu** [comment]
/assign @klueska
Would you mind taking a look?

File: cmd/kubelet/app/server.go
Comment (pacoxu) [reply cmd/kubelet/app/server.go:258]: /sig instrumentation
/cc @serathius
I am afraid that there would be some similar problems in other places. We should take care.

**Review by pacoxu — APPROVED** [review]

File: cmd/kubelet/app/server.go
Comment (serathius) [reply cmd/kubelet/app/server.go:258]: cc @pohly

**Comment by serathius** [comment]
This is not following the current Structured Logging guidelines.
/hold
/assign @pohly

File: cmd/kubelet/app/server.go
Comment (pohly) [reply cmd/kubelet/app/server.go:258]: For values where structured logging (i.e. dumping it as a struct in this case) is never going to be needed, then I don't see a problem with converting to a string at the call site.

But I agree that this looks like a workaround. `KubeletConfiguration` has a `String` implementation and that implementation is broken.

**Comment by SataQiu** [comment]
> This is not following the current Structured Logging guidelines. /hold /assign @pohly

Is there a best practice about this? 
Is there a helper function for printing detailed data structure in `klog.InfoS`?

**Comment by serathius** [comment]
> > This is not following the current Structured Logging guidelines. /hold /assign @pohly
> 
> Is there a best practice about this? Is there a helper function for printing detailed data structure in `klog.InfoS`?

deferring to @pohly

**Comment by pohly** [comment]
The best practice is to implement `fmt.Stringer` (= `String() string`)  and `logr.LogMarshaler` (= `MarshalLog() interface{}`).

In this case a dump of the entire struct doesn't seem necessary, but then at least fix `String()` so that it actually includes the entire struct.

**Comment by SataQiu** [comment]
> The best practice is to implement `fmt.Stringer` (= `String() string`) and `logr.LogMarshaler` (= `MarshalLog() interface{}`).
> 
> In this case a dump of the entire struct doesn't seem necessary, but then at least fix `String()` so that it actually includes the entire struct.

@pohly  I found that `klog` will use `String()` first if both the `fmt.Stringer` and `logr.LogMarshaler` interfaces are implemented. I think `logr.LogMarshaler` should be preferred, but it is not the current behavior.
https://github.com/kubernetes/kubernetes/blob/8e01ee79bf78aee1f6a42443acdfc33c89c09952/vendor/k8s.io/klog/v2/internal/serialize/keyvalues.go#L195-L204

So, I'd like to implement only the `logr.LogMarshaler` interface to support logging here. WDYT?

**Comment by pohly** [comment]
> I found that klog will use String() first if both the fmt.Stringer and logr.LogMarshaler interfaces are implemented.

Correct. For a type that supports both, JSON output will prefer `logr.LogMarshaler`. Text output will prefer `fmt.Stringer`. That enables a type to support pretty-printing for text output and providing a struct for structured output.

> I think logr.LogMarshaler should be preferred

Why?

**Comment by pohly** [comment]
> The test results show that the KubeletConfiguration can be output correctly after the implementation of interface logr.LogMarshaler.

Implementing `logr.LogMarshaler` doesn't fix the underlying bug: `KubeletConfiguration.String` doesn't serialize the struct.

**Comment by SataQiu** [comment]
> For a type that supports both, JSON output will prefer `logr.LogMarshaler`. Text output will prefer `fmt.Stringer`.

Thanks @pohly In this case, which format should we choose?  JSON or Text?

**Comment by pohly** [comment]
> In this case, which format should we choose? JSON or Text？

Implement both (`fmt.Stringer` and `logr.LogMarshaler`), then you get the right output depending on which logging format is selected.

**Comment by pacoxu** [comment]
/unhold
/lgtm
/cc @pohly

would you take a look again?

**Review by pohly — APPROVED** [review]
/lgtm

**Comment by SataQiu** [comment]
/assign @liggitt 
for api approve (from bot suggestion)

**Comment by liggitt** [comment]
I don't think we should add String() implementations to random API types for logging purposes. If we want a lazy-evaluated YAML marshaler that wraps a runtime.Object for logging purposes, let's make that and wrap the kubelet configuration in it when we call `klog.V(5).InfoS("KubeletConfiguration", "configuration", config)`

**Comment by liggitt** [comment]
hopefully people adding log lines check what they look like before committing/merging

**Comment by SataQiu** [comment]
+1 for wrapper approach and logcheck lint

**Comment by pohly** [comment]
@SataQiu would a wrapper that formats as single-line JSON be okay? That could be added klog.

After thinking about it a bit more, I don't think that component-base/logs would be a good place. component-base is huge and targeted towards the main binaries, so using it in some small library package wouldn't be possible.

**Comment by SataQiu** [comment]
> would a wrapper that formats as single-line JSON be okay? 

I think that is OK, at least can solve this problem. And we can support more formats in the future if we need to.

> That could be added to klog

+1 for klog, and we must be careful to avoid import cycle.

**Comment by pohly** [comment]
In related news: klog now falls back to JSON, which makes, for example, this log line more useful:
```
Apr 29 19:08:08 kind-worker kubelet[294]: I0429 19:08:08.264948     294 container_manager_linux.go:271] "Creating Container Manager object based on Node Config" nodeConfig={"RuntimeCgroupsName":"","SystemCgroupsName":"","KubeletCgroupsName":"","KubeletOOMScoreAdj":-999,"ContainerRuntime":"","CgroupsPerQOS":true,"CgroupRoot":"/kubelet","CgroupDriver":"systemd","KubeletRootDir":"/var/lib/kubelet","ProtectKernelDefaults":false,"KubeReservedCgroupName":"","SystemReservedCgroupName":"","ReservedSystemCPUs":{},"EnforceNodeAllocatable":{"pods":{}},"KubeReserved":null,"SystemReserved":null,"HardEvictionThresholds":[],"QOSReserved":{},"CPUManagerPolicy":"none","CPUManagerPolicyOptions":null,"TopologyManagerScope":"container","CPUManagerReconcilePeriod":10000000000,"ExperimentalMemoryManagerPolicy":"None","ExperimentalMemoryManagerReservedMemory":null,"PodPidsLimit":-1,"EnforceCPULimits":true,"CPUCFSQuotaPeriod":100000000,"TopologyManagerPolicy":"none","ExperimentalTopologyManagerPolicyOptions":null}
```

Previously that was:
```
Apr 28 14:42:50 kind-worker kubelet[293]: I0428 14:42:50.965399     293 container_manager_linux.go:271] "Creating Container Manager object based on Node Config" nodeConfig={RuntimeCgroupsName: SystemCgroupsName: KubeletCgroupsName: KubeletOOMScoreAdj:-999 ContainerRuntime: CgroupsPerQOS:true CgroupRoot:/kubelet CgroupDriver:systemd KubeletRootDir:/var/lib/kubelet ProtectKernelDefaults:false NodeAllocatableConfig:{KubeReservedCgroupName: SystemReservedCgroupName: ReservedSystemCPUs: EnforceNodeAllocatable:map[pods:{}] KubeReserved:map[] SystemReserved:map[] HardEvictionThresholds:[]} QOSReserved:map[] CPUManagerPolicy:none CPUManagerPolicyOptions:map[] TopologyManagerScope:container CPUManagerReconcilePeriod:10s ExperimentalMemoryManagerPolicy:None ExperimentalMemoryManagerReservedMemory:[] PodPidsLimit:-1 EnforceCPULimits:true CPUCFSQuotaPeriod:100ms TopologyManagerPolicy:none ExperimentalTopologyManagerPolicyOptions:map[]}
```

This could be even better with `omitempty`.

**Comment by SataQiu** [comment]
@pohly Updated to use `klog.Format`. It works fine :)
@liggitt PTAL again.

And a follow PR #117907 was submitted for `AttachedVolume`.

**Pull request merged**
